src/libsyntax/ast.rs

```diff
@@ -2106,20 +2106,28 @@ pub struct AttrItem {
 }
 
 /// Metadata associated with an item.
-/// Doc-comments are promoted to attributes that have `is_sugared_doc = true`.
 #[derive(Clone, RustcEncodable, RustcDecodable, Debug)]
 pub struct Attribute {
-    pub item: AttrItem,
+    pub kind: AttrKind,
     pub id: AttrId,
     pub style: AttrStyle,
-    pub is_sugared_doc: bool,
     pub span: Span,
 }
 
-// Compatibility impl to avoid churn, consider removing.
-impl std::ops::Deref for Attribute {
-    type Target = AttrItem;
-    fn deref(&self) -> &Self::Target { &self.item }
+#[derive(Clone, RustcEncodable, RustcDecodable, Debug)]
+pub enum AttrKind {
+    /// A normal attribute. Unsugared doc comments (e.g. `#[doc="..."]`) are
+    /// represented with this variant.
+    Normal(AttrItem),
+
+    /// A sugared doc comment attribute (e.g. `/// ...`). Note that this is a
+    /// more compact and thus cheaper representation than unsugared doc
+    /// comments.
```

Comment:
I feel we could be more positive about this; it's not just an optimization hack but more "correct" from the POV of being syntactically distinct (which is relevant at the level of the concrete syntax tree).